extensions/mutiny/runtime/src/main/java/io/quarkus/mutiny/runtime/MutinyInfrastructure.java

```diff
@@ -1,42 +1,144 @@
 package io.quarkus.mutiny.runtime;
 
-import java.util.concurrent.Executor;
-import java.util.concurrent.ExecutorService;
-import java.util.concurrent.RejectedExecutionException;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.*;
 import java.util.function.BooleanSupplier;
 import java.util.function.Consumer;
 
 import org.jboss.logging.Logger;
+import org.jboss.threads.ContextHandler;
 
 import io.quarkus.runtime.ShutdownContext;
 import io.quarkus.runtime.annotations.Recorder;
 import io.smallrye.mutiny.infrastructure.Infrastructure;
+import io.smallrye.mutiny.infrastructure.MutinyScheduler;
 
 @Recorder
 public class MutinyInfrastructure {
 
     public static final String VERTX_EVENT_LOOP_THREAD_PREFIX = "vert.x-eventloop-thread-";
 
-    public void configureMutinyInfrastructure(ExecutorService exec, ShutdownContext shutdownContext) {
-        //mutiny leaks a ScheduledExecutorService if you don't do this
+    public void configureMutinyInfrastructure(ExecutorService executor, ShutdownContext shutdownContext,
+            ContextHandler<Object> contextHandler) {
+        // Mutiny leaks a ScheduledExecutorService if we don't do this
         Infrastructure.getDefaultWorkerPool().shutdown();
-        Infrastructure.setDefaultExecutor(new Executor() {
+
+        // Since executor is not a ScheduledExecutorService and Mutiny needs one for scheduling we have to adapt one around the provided executor
+        MutinyScheduler mutinyScheduler = new MutinyScheduler(executor) {
+            @Override
+            protected <V> RunnableScheduledFuture<V> decorateTask(Runnable runnable, RunnableScheduledFuture<V> task) {
+                return super.decorateTask(runnable,
+                        new ContextualRunnableScheduledFuture<>(contextHandler, contextHandler.captureContext(), task));
```

Comment:
Wouldn't `contextHandler.captureContext()` fail if contextHandler is not set (null)? Maybe the switch of contextHandler != null ? new MutinyScheduler(executor) { ... } :  new MutinyScheduler()` and then in the `ContextualRunnableScheduledFuture` you wouldn't need to check if contextHandler is null or not, since it will only be used if it is not null... Maybe that approach is slightly better?

Reply:
Ah good catch. I'd rather avoid having 2 classes (plain `MutinyScheduler` vs an anonymous extension. Null checks are cheap, it won't be a performance hog.